src/Tests/Microsoft.NET.TestFramework/TestAssetsManager.cs

```diff
@@ -124,6 +125,12 @@ public string GetAndValidateTestProjectDirectory(string testProjectName, string
                 directoryName = directoryName.Append(testProjectName);
             }
 
+            if (unique)
+            {
+                // avoid collisions between derived test classes using same base method concurrently
+                directoryName = directoryName.Append(Guid.NewGuid().ToString("N").Substring(0, 8));
```

Comment:
We've discussed doing this in the past and decided against it, as we don't want the number of test dirs to constantly grow when a dev runs tests multiple times. If you're running into test dir collisions on theories, you can use the identifier param.